File: src/main/java/in/costea/wiles/commands/CodeBlockCommand.java
package in.costea.wiles.commands;

import in.costea.wiles.data.CompilationExceptionsCollection;
import in.costea.wiles.enums.WhenRemoveToken;
import in.costea.wiles.exceptions.AbstractCompilationException;
import in.costea.wiles.exceptions.UnexpectedTokenException;
import in.costea.wiles.services.TokenTransmitter;
import in.costea.wiles.statics.Constants;

import java.util.ArrayList;
import java.util.List;

import static in.costea.wiles.builders.ExpectParamsBuilder.*;
import static in.costea.wiles.commands.ExpressionCommand.*;
import static in.costea.wiles.statics.Constants.*;

public class CodeBlockCommand extends AbstractCommand
{
    private final List<ExpressionCommand> components = new ArrayList<>();
    private final CompilationExceptionsCollection exceptions = new CompilationExceptionsCollection();
    private final boolean standAlone;

    public CodeBlockCommand(TokenTransmitter transmitter, boolean standAlone)
    {
        super(transmitter);
        this.standAlone = standAlone;
    }

    @Override
    public Constants.SYNTAX_TYPE getType()
    {
        return Constants.SYNTAX_TYPE.CODE_BLOCK;
    }

    @Override
    public List<ExpressionCommand> getComponents()
    {
        return components;
    }

    private void readOneStatement() throws AbstractCompilationException {
        if (transmitter.expectMaybe(tokenOf(isContainedIn(STATEMENT_TERMINATORS)).dontIgnoreNewLine()).isPresent())
            return;
        ExpressionCommand expressionCommand;
        boolean innerExpression = transmitter.expectMaybe(tokenOf(ROUND_BRACKET_START_ID)).isPresent();

        var optionalToken=transmitter.expectMaybe(tokenOf(isContainedIn(UNARY_OPERATORS)).or(IS_LITERAL));
        if (optionalToken.isPresent())
            expressionCommand = new ExpressionCommand(optionalToken.get(), transmitter, innerExpression?INSIDE_ROUND:REGULAR);
        else
        {
            optionalToken=transmitter.expectMaybe(tokenOf(DECLARE_METHOD_ID));
            if (standAlone && optionalToken.isPresent())
                throw new UnexpectedTokenException("Cannot declare method in body-only mode!", optionalToken.get().location());
            else
            {
                var token=transmitter.expect(tokenOf(ANYTHING));
                throw new UnexpectedTokenException(TOKENS_INVERSE.get(token.content()), token.location());
            }
        }

        CompilationExceptionsCollection newExceptions = expressionCommand.process();
        if(newExceptions.size()>0)
            throw newExceptions.get(0);

        components.add(expressionCommand);
    }

    @Override
    public CompilationExceptionsCollection process()
    {
        try
        {
            if (!standAlone && transmitter.expectMaybe(tokenOf(DO_ID)).isPresent())
            {
                if(transmitter.expectMaybe(tokenOf(NOTHING_ID)).isEmpty())
                {
                    readOneStatement();
                }
                return exceptions;
            }
            else
            {
                if(!standAlone)
                    transmitter.expect(tokenOf(START_BLOCK_ID));
                while(!transmitter.tokensExhausted())
                {
                    if(!standAlone && transmitter.expectMaybe(tokenOf(END_BLOCK_ID).removeTokenWhen(WhenRemoveToken.Never)).isPresent())
                        break;
                    readOneStatement();
                }
                if(!standAlone) transmitter.expect(tokenOf(END_BLOCK_ID));
            }
        }
        catch (AbstractCompilationException ex)
        {
            exceptions.add(ex);
        }
        return exceptions;
    }
}

File: src/main/java/in/costea/wiles/commands/CodeBlockCommand.kt
package `in`.costea.wiles.commands

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import `in`.costea.wiles.builders.CommandFactory
import `in`.costea.wiles.builders.ExpectParamsBuilder.Companion.tokenOf
import `in`.costea.wiles.commands.expressions.AssignableExpressionCommand
import `in`.costea.wiles.data.CompilationExceptionsCollection
import `in`.costea.wiles.enums.SyntaxType
import `in`.costea.wiles.enums.WhenRemoveToken
import `in`.costea.wiles.exceptions.AbstractCompilationException
import `in`.costea.wiles.services.TokenTransmitter
import `in`.costea.wiles.statics.Constants.DO_ID
import `in`.costea.wiles.statics.Constants.END_BLOCK_ID
import `in`.costea.wiles.statics.Constants.EXPECT_TERMINATOR
import `in`.costea.wiles.statics.Constants.NOTHING_ID
import `in`.costea.wiles.statics.Constants.START_BLOCK_ID

class CodeBlockCommand(transmitter: TokenTransmitter, private val outerMost: Boolean) : AbstractCommand(transmitter) {
    private val components: MutableList<AbstractCommand> = ArrayList()
    private val exceptions: CompilationExceptionsCollection = CompilationExceptionsCollection()

    @JsonProperty
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private var compiledSuccessfully: Boolean? = null

    override val type: SyntaxType
        get() = SyntaxType.CODE_BLOCK

    override fun getComponents(): List<AbstractCommand> {
        return components
    }

    @Throws(AbstractCompilationException::class)
    private fun readOneStatement() {
        if (transmitter.expectMaybe(EXPECT_TERMINATOR).isPresent) return

        val command: AbstractCommand = CommandFactory(transmitter)
            .of(AssignableExpressionCommand::class.java)
            .of(DeclarationCommand::class.java)
            .create()

        val newExceptions = command.process()
        if (newExceptions.size > 0) throw newExceptions[0]
        components.add(command)
    }

    override fun process(): CompilationExceptionsCollection {
        try {
            if (!outerMost && transmitter.expectMaybe(tokenOf(DO_ID)).isPresent) {
                if (transmitter.expectMaybe(tokenOf(NOTHING_ID)).isEmpty)
                    readOneStatement()
            } else {
                if (!outerMost) transmitter.expect(tokenOf(START_BLOCK_ID))
                while (!transmitter.tokensExhausted()) {
                    if (!outerMost && transmitter.expectMaybe(tokenOf(END_BLOCK_ID).removeWhen(WhenRemoveToken.Never))
                            .isPresent)
                        break
                    readOneStatement()
                }
                if (!outerMost) transmitter.expect(tokenOf(END_BLOCK_ID))
            }
        } catch (ex: AbstractCompilationException) {
            exceptions.add(ex)
        }
        compiledSuccessfully = exceptions.isEmpty()
        return exceptions
    }
}